plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/contributors/FirDeclarationFromUnresolvedNameContributor.kt

```diff
@@ -0,0 +1,142 @@
+        currentDeclarationInFakeFile: KtNamedDeclaration
+    ): Boolean {
+        val refExprParent = unresolvedRef.parent
+        val refExprIsCallRef = refExprParent is KtCallableReferenceExpression
+        val refExprIsCall = refExprParent is KtCallExpression
+        val refExprIsType = refExprParent is KtUserType
+        val receiver = if (refExprIsCall) refExprParent.getReceiverForSelector() else unresolvedRef.getReceiverForSelector()
+        return when (val symbol = currentDeclarationInFakeFile.getSymbol()) {
+            is KtCallableSymbol -> {
+                if (refExprIsType) return false
+                // If current declaration is a function, we only offer names of unresolved function calls. For property declarations, we
+                // always offer an unresolved usage because a property may be called if the object has an `invoke` method.
+                if (!refExprIsCallRef && !refExprIsCall && symbol is KtFunctionLikeSymbol) return false
+                if (receiver != null) {
+                    val actualReceiverType = receiver.getKtType() ?: return false
+                    val expectedReceiverType = getReceiverType(symbol) ?: return false
+                    actualReceiverType isSubTypeOf expectedReceiverType
+                } else {
+                    // If there is no explicit receiver at call-site, we check if any implicit receiver at call-site matches the extension
+                    // receiver type for the current declared symbol
+                    val extensionReceiverType = symbol.receiverType?.type ?: return true
+                    getImplicitReceiverTypesAtPosition(unresolvedRef).any { it isSubTypeOf extensionReceiverType }
+                }
+            }
+            is KtClassOrObjectSymbol -> {
+                if (receiver != null) return false
+                if (refExprIsType) return true
+                if (refExprIsCall) {
+                    symbol.classKind == KtClassKind.CLASS
+                } else {
+                    !(symbol.classKind != KtClassKind.OBJECT && symbol.classKind != KtClassKind.ENUM_ENTRY)
+                }
+            }
+            else -> false
+        }
+    }
+
+    private fun PsiElement.getReceiverForSelector(): KtExpression? {
+        val qualifiedExpression = (parent as? KtDotQualifiedExpression)?.takeIf { it.selectorExpression == this } ?: return null
+        return qualifiedExpression.receiverExpression
+    }
+
+    private fun getReceiverType(symbol: KtCallableSymbol): KtType? {
+        return symbol.receiverType?.type ?: (symbol as? KtPossibleMemberSymbol)?.dispatchType
+    }
+
+
+    private fun PsiElement.getCurrentDeclarationAtCaret(): KtNamedDeclaration? {
+        return when (val parent = parent) {
+            is KtNamedDeclaration -> parent
+            is KtNameReferenceExpression -> {
+                // In a fake completion file, a property or function under construction is appended with placeholder text "X.f$", which is
```

Comment:
I suggest you to reference the code that selects the dummy identifier in the completion here, so the reader won't have to look it up by himself - it can be something like `see <method> for the details`